File: ios/EditorTheme.swift
import UIKit

struct EditorTextStyle {
    var fontFamily: String?
    var fontSize: CGFloat?
    var fontWeight: String?
    var fontStyle: String?
    var color: UIColor?
    var lineHeight: CGFloat?
    var spacingAfter: CGFloat?

    init(
        fontFamily: String? = nil,
        fontSize: CGFloat? = nil,
        fontWeight: String? = nil,
        fontStyle: String? = nil,
        color: UIColor? = nil,
        lineHeight: CGFloat? = nil,
        spacingAfter: CGFloat? = nil
    ) {
        self.fontFamily = fontFamily
        self.fontSize = fontSize
        self.fontWeight = fontWeight
        self.fontStyle = fontStyle
        self.color = color
        self.lineHeight = lineHeight
        self.spacingAfter = spacingAfter
    }

    init(dictionary: [String: Any]) {
        fontFamily = dictionary["fontFamily"] as? String
        fontSize = EditorTheme.cgFloat(dictionary["fontSize"])
        fontWeight = dictionary["fontWeight"] as? String
        fontStyle = dictionary["fontStyle"] as? String
        color = EditorTheme.color(from: dictionary["color"])
        lineHeight = EditorTheme.cgFloat(dictionary["lineHeight"])
        spacingAfter = EditorTheme.cgFloat(dictionary["spacingAfter"])
    }

    func merged(with override: EditorTextStyle?) -> EditorTextStyle {
        guard let override else { return self }
        return EditorTextStyle(
            fontFamily: override.fontFamily ?? fontFamily,
            fontSize: override.fontSize ?? fontSize,
            fontWeight: override.fontWeight ?? fontWeight,
            fontStyle: override.fontStyle ?? fontStyle,
            color: override.color ?? color,
            lineHeight: override.lineHeight ?? lineHeight,
            spacingAfter: override.spacingAfter ?? spacingAfter
        )
    }

    func resolvedFont(fallback: UIFont) -> UIFont {
        let size = fontSize ?? fallback.pointSize
        var font = fallback.withSize(size)

        if let fontFamily,
           let familyFont = UIFont(name: fontFamily, size: size) {
            font = familyFont
        } else if let fontWeight {
            font = UIFont.systemFont(ofSize: size, weight: EditorTheme.fontWeight(from: fontWeight))
        }

        var traits = font.fontDescriptor.symbolicTraits
        if EditorTheme.shouldApplyBoldTrait(fontWeight) {
            traits.insert(.traitBold)
        }
        if fontStyle == "italic" {
            traits.insert(.traitItalic)
        }

        if traits != font.fontDescriptor.symbolicTraits,
           let descriptor = font.fontDescriptor.withSymbolicTraits(traits) {
            font = UIFont(descriptor: descriptor, size: size)
        }

        return font
    }
}

struct EditorListTheme {
    var indent: CGFloat?
    var itemSpacing: CGFloat?
    var markerColor: UIColor?
    var markerScale: CGFloat?

    init(dictionary: [String: Any]) {
        indent = EditorTheme.cgFloat(dictionary["indent"])
        itemSpacing = EditorTheme.cgFloat(dictionary["itemSpacing"])
        markerColor = EditorTheme.color(from: dictionary["markerColor"])
        markerScale = EditorTheme.cgFloat(dictionary["markerScale"])
    }
}

struct EditorHorizontalRuleTheme {
    var color: UIColor?
    var thickness: CGFloat?
    var verticalMargin: CGFloat?

    init(dictionary: [String: Any]) {
        color = EditorTheme.color(from: dictionary["color"])
        thickness = EditorTheme.cgFloat(dictionary["thickness"])
        verticalMargin = EditorTheme.cgFloat(dictionary["verticalMargin"])
    }
}

struct EditorMentionTheme {
    var textColor: UIColor?
    var backgroundColor: UIColor?
    var borderColor: UIColor?
    var borderWidth: CGFloat?
    var borderRadius: CGFloat?
    var fontWeight: String?
    var popoverBackgroundColor: UIColor?
    var popoverBorderColor: UIColor?
    var popoverBorderWidth: CGFloat?
    var popoverBorderRadius: CGFloat?
    var popoverShadowColor: UIColor?
    var optionTextColor: UIColor?
    var optionSecondaryTextColor: UIColor?
    var optionHighlightedBackgroundColor: UIColor?
    var optionHighlightedTextColor: UIColor?

    init(dictionary: [String: Any]) {
        textColor = EditorTheme.color(from: dictionary["textColor"])
        backgroundColor = EditorTheme.color(from: dictionary["backgroundColor"])
        borderColor = EditorTheme.color(from: dictionary["borderColor"])
        borderWidth = EditorTheme.cgFloat(dictionary["borderWidth"])
        borderRadius = EditorTheme.cgFloat(dictionary["borderRadius"])
        fontWeight = dictionary["fontWeight"] as? String
        popoverBackgroundColor = EditorTheme.color(from: dictionary["popoverBackgroundColor"])
        popoverBorderColor = EditorTheme.color(from: dictionary["popoverBorderColor"])
        popoverBorderWidth = EditorTheme.cgFloat(dictionary["popoverBorderWidth"])
        popoverBorderRadius = EditorTheme.cgFloat(dictionary["popoverBorderRadius"])
        popoverShadowColor = EditorTheme.color(from: dictionary["popoverShadowColor"])
        optionTextColor = EditorTheme.color(from: dictionary["optionTextColor"])
        optionSecondaryTextColor = EditorTheme.color(from: dictionary["optionSecondaryTextColor"])
        optionHighlightedBackgroundColor = EditorTheme.color(from: dictionary["optionHighlightedBackgroundColor"])
        optionHighlightedTextColor = EditorTheme.color(from: dictionary["optionHighlightedTextColor"])
    }
}

struct EditorToolbarTheme {
    var backgroundColor: UIColor?
    var borderColor: UIColor?
    var borderWidth: CGFloat?
    var borderRadius: CGFloat?
    var keyboardOffset: CGFloat?
    var horizontalInset: CGFloat?
    var separatorColor: UIColor?
    var buttonColor: UIColor?
    var buttonActiveColor: UIColor?
    var buttonDisabledColor: UIColor?
    var buttonActiveBackgroundColor: UIColor?
    var buttonBorderRadius: CGFloat?

    init(dictionary: [String: Any]) {
        backgroundColor = EditorTheme.color(from: dictionary["backgroundColor"])
        borderColor = EditorTheme.color(from: dictionary["borderColor"])
        borderWidth = EditorTheme.cgFloat(dictionary["borderWidth"])
        borderRadius = EditorTheme.cgFloat(dictionary["borderRadius"])
        keyboardOffset = EditorTheme.cgFloat(dictionary["keyboardOffset"])
        horizontalInset = EditorTheme.cgFloat(dictionary["horizontalInset"])
        separatorColor = EditorTheme.color(from: dictionary["separatorColor"])
        buttonColor = EditorTheme.color(from: dictionary["buttonColor"])
        buttonActiveColor = EditorTheme.color(from: dictionary["buttonActiveColor"])
        buttonDisabledColor = EditorTheme.color(from: dictionary["buttonDisabledColor"])
        buttonActiveBackgroundColor = EditorTheme.color(from: dictionary["buttonActiveBackgroundColor"])
        buttonBorderRadius = EditorTheme.cgFloat(dictionary["buttonBorderRadius"])
    }
}

struct EditorContentInsets {
    var top: CGFloat?
    var right: CGFloat?
    var bottom: CGFloat?
    var left: CGFloat?

    init(dictionary: [String: Any]) {
        top = EditorTheme.cgFloat(dictionary["top"])
        right = EditorTheme.cgFloat(dictionary["right"])
        bottom = EditorTheme.cgFloat(dictionary["bottom"])
        left = EditorTheme.cgFloat(dictionary["left"])
    }
}

struct EditorTheme {
    var text: EditorTextStyle?
    var paragraph: EditorTextStyle?
    var headings: [String: EditorTextStyle] = [:]
    var list: EditorListTheme?
    var horizontalRule: EditorHorizontalRuleTheme?
    var mentions: EditorMentionTheme?
    var toolbar: EditorToolbarTheme?
    var backgroundColor: UIColor?
    var borderRadius: CGFloat?
    var contentInsets: EditorContentInsets?

    static func from(json: String?) -> EditorTheme? {
        guard let json, !json.isEmpty,
              let data = json.data(using: .utf8),
              let raw = try? JSONSerialization.jsonObject(with: data) as? [String: Any]
        else {
            return nil
        }
        return EditorTheme(dictionary: raw)
    }

    init(dictionary: [String: Any]) {
        if let text = dictionary["text"] as? [String: Any] {
            self.text = EditorTextStyle(dictionary: text)
        }
        if let paragraph = dictionary["paragraph"] as? [String: Any] {
            self.paragraph = EditorTextStyle(dictionary: paragraph)
        }
        if let headings = dictionary["headings"] as? [String: Any] {
            for level in ["h1", "h2", "h3", "h4", "h5", "h6"] {
                if let style = headings[level] as? [String: Any] {
                    self.headings[level] = EditorTextStyle(dictionary: style)
                }
            }
        }
        if let list = dictionary["list"] as? [String: Any] {
            self.list = EditorListTheme(dictionary: list)
        }
        if let horizontalRule = dictionary["horizontalRule"] as? [String: Any] {
            self.horizontalRule = EditorHorizontalRuleTheme(dictionary: horizontalRule)
        }
        if let mentions = dictionary["mentions"] as? [String: Any] {
            self.mentions = EditorMentionTheme(dictionary: mentions)
        }
        if let toolbar = dictionary["toolbar"] as? [String: Any] {
            self.toolbar = EditorToolbarTheme(dictionary: toolbar)
        }
        backgroundColor = EditorTheme.color(from: dictionary["backgroundColor"])
        borderRadius = EditorTheme.cgFloat(dictionary["borderRadius"])
        if let contentInsets = dictionary["contentInsets"] as? [String: Any] {
            self.contentInsets = EditorContentInsets(dictionary: contentInsets)
        }
    }

    func effectiveTextStyle(for nodeType: String) -> EditorTextStyle {
        var style = text ?? EditorTextStyle()
        if nodeType == "paragraph" {
            style = style.merged(with: paragraph)
            if paragraph?.lineHeight == nil {
                style.lineHeight = nil
            }
        }
        style = style.merged(with: headings[nodeType])
        return style
    }

    static func cgFloat(_ value: Any?) -> CGFloat? {
        guard let number = value as? NSNumber else { return nil }
        return CGFloat(truncating: number)
    }

    static func fontWeight(from value: String) -> UIFont.Weight {
        switch value {
        case "100": return .ultraLight
        case "200": return .thin
        case "300": return .light
        case "500": return .medium
        case "600": return .semibold
        case "700", "bold": return .bold
        case "800": return .heavy
        case "900": return .black
        default: return .regular
        }
    }

    static func shouldApplyBoldTrait(_ value: String?) -> Bool {
        guard let value else { return false }
        return value == "bold" || Int(value).map { $0 >= 600 } == true
    }

    static func color(from value: Any?) -> UIColor? {
        guard let raw = value as? String else { return nil }
        let string = raw.trimmingCharacters(in: .whitespacesAndNewlines).lowercased()

        if let hexColor = colorFromHex(string) {
            return hexColor
        }
        if let rgbColor = colorFromRGBFunction(string) {
            return rgbColor
        }

        switch string {
        case "black": return .black
        case "white": return .white
        case "red": return .red
        case "green": return .green
        case "blue": return .blue
        case "gray", "grey": return .gray
        case "clear", "transparent": return .clear
        default: return nil
        }
    }

    private static func colorFromHex(_ string: String) -> UIColor? {
        guard string.hasPrefix("#") else { return nil }
        let hex = String(string.dropFirst())

        switch hex.count {
        case 3:
            let chars = Array(hex)
            return UIColor(
                red: component(String(repeating: String(chars[0]), count: 2)),
                green: component(String(repeating: String(chars[1]), count: 2)),
                blue: component(String(repeating: String(chars[2]), count: 2)),
                alpha: 1
            )
        case 4:
            let chars = Array(hex)
            return UIColor(
                red: component(String(repeating: String(chars[0]), count: 2)),
                green: component(String(repeating: String(chars[1]), count: 2)),
                blue: component(String(repeating: String(chars[2]), count: 2)),
                alpha: component(String(repeating: String(chars[3]), count: 2))
            )
        case 6:
            return UIColor(
                red: component(String(hex.prefix(2))),
                green: component(String(hex.dropFirst(2).prefix(2))),
                blue: component(String(hex.dropFirst(4).prefix(2))),
                alpha: 1
            )
        case 8:
            return UIColor(
                red: component(String(hex.prefix(2))),
                green: component(String(hex.dropFirst(2).prefix(2))),
                blue: component(String(hex.dropFirst(4).prefix(2))),
                alpha: component(String(hex.dropFirst(6).prefix(2)))
            )
        default:
            return nil
        }
    }

    private static func colorFromRGBFunction(_ string: String) -> UIColor? {
        let isRGBA = string.hasPrefix("rgba(") && string.hasSuffix(")")
        let isRGB = string.hasPrefix("rgb(") && string.hasSuffix(")")
        guard isRGBA || isRGB else { return nil }

        let start = string.index(string.startIndex, offsetBy: isRGBA ? 5 : 4)
        let end = string.index(before: string.endIndex)
        let parts = string[start..<end]
            .split(separator: ",")
            .map { $0.trimmingCharacters(in: .whitespacesAndNewlines) }

        guard parts.count == (isRGBA ? 4 : 3),
              let red = Double(parts[0]),
              let green = Double(parts[1]),
              let blue = Double(parts[2])
        else {
            return nil
        }

        let alpha = isRGBA ? (Double(parts[3]) ?? 1) : 1
        return UIColor(
            red: red / 255,
            green: green / 255,
            blue: blue / 255,
            alpha: alpha
        )
    }

    private static func component(_ hex: String) -> CGFloat {
        CGFloat(Int(hex, radix: 16) ?? 0) / 255
    }
}

File: src/EditorTheme.ts
/**
 * Font weight accepted by every themed text style. Numeric weights are
 * strings, matching React Native's `fontWeight`.
 */
export type EditorFontWeight =
    | 'normal'
    | 'bold'
    | '100'
    | '200'
    | '300'
    | '400'
    | '500'
    | '600'
    | '700'
    | '800'
    | '900';

/** Font slant accepted by every themed text style. */
export type EditorFontStyle = 'normal' | 'italic';

/** The mention itself, rendered inline in the document. */
export interface EditorMentionNodeTheme {
    /** Color of the mention label. */
    textColor?: string;
    /** Fill drawn behind the mention label. */
    backgroundColor?: string;
    /** Weight of the mention label. */
    fontWeight?: EditorFontWeight;
    /** Drawn by `NativeProseViewer`. The editor renders the node with text
     *  spans and cannot stroke a border, so these are ignored there. */
    borderColor?: string;
    borderWidth?: number;
    borderRadius?: number;
}

/** A single row in the suggestion list. */
export interface EditorMentionSuggestionOptionTheme {
    /** Color of the row's primary text (`MentionSuggestion.title`). */
    textColor?: string;
    /** Color of the row's secondary text (`MentionSuggestion.subtitle`). */
    secondaryTextColor?: string;
    /** Fill drawn behind an unhighlighted row. */
    backgroundColor?: string;
    borderColor?: string;
    /** Row border width, in layout units. */
    borderWidth?: number;
    /** Row corner radius, in layout units. */
    borderRadius?: number;
    /** Weight of the row's primary text. */
    fontWeight?: EditorFontWeight;
    /** Fill drawn behind the row under the cursor or pointer. */
    highlightedBackgroundColor?: string;
    /** Primary text color for the row under the cursor or pointer. */
    highlightedTextColor?: string;
}

/** The container behind the suggestion rows. Honored by the JavaScript
 *  `EditorToolbar`; the native keyboard toolbar styles rows only. */
export interface EditorMentionSuggestionsTheme {
    backgroundColor?: string;
    borderColor?: string;
    /** Container border width, in layout units. */
    borderWidth?: number;
    /** Container corner radius, in layout units. */
    borderRadius?: number;
    shadowColor?: string;
    /** Styling for each suggestion row inside the container. */
    option?: EditorMentionSuggestionOptionTheme;
}

/**
 * Mention styling. It is configured on the mentions addon
 * (`EditorAddons.mentions.theme` for the editor,
 * `NativeProseViewerAddons.mentions.theme` for the viewer) rather than on
 * {@link EditorTheme}, and can be overridden per mention through
 * `MentionsAddonConfig.resolveTheme`.
 */
export interface EditorMentionTheme {
    node?: EditorMentionNodeTheme;
    suggestions?: EditorMentionSuggestionsTheme;
}

/**
 * Typography for one kind of text block. Unset fields fall back to the
 * enclosing style — see the cascade described on {@link EditorTheme}.
 */
export interface EditorTextStyle {
    /** Platform font family name. Falls back to the system font when unresolvable. */
    fontFamily?: string;
    /** Text size, in layout units. */
    fontSize?: number;
    fontWeight?: EditorFontWeight;
    fontStyle?: EditorFontStyle;
    /** Text color. */
    color?: string;
    /** Total line height, in layout units — not a multiplier of `fontSize`. */
    lineHeight?: number;
    /** Space below each block using this style, in layout units. */
    spacingAfter?: number;
}

/** Styling for text carrying the `link` mark. */
export interface EditorLinkTheme {
    fontFamily?: string;
    /** Text size, in layout units. */
    fontSize?: number;
    fontWeight?: EditorFontWeight;
    fontStyle?: EditorFontStyle;
    color?: string;
    /** Fill drawn behind the link text. */
    backgroundColor?: string;
    /** Whether link text is underlined. */
    underline?: boolean;
}

/** Per-level heading typography, merged over the base `text` style. */
export interface EditorHeadingTheme {
    h1?: EditorTextStyle;
    h2?: EditorTextStyle;
    h3?: EditorTextStyle;
    h4?: EditorTextStyle;
    h5?: EditorTextStyle;
    h6?: EditorTextStyle;
}

export type EditorOrderedListNumberingScheme =
    | 'decimal'
    | 'lowerAlpha'
    | 'upperAlpha'
    | 'lowerRoman'
    | 'upperRoman';

export interface EditorOrderedListMarkerTheme {
    /**
     * Numbering schemes selected by visual list depth and cycled when necessary.
     * Defaults to `decimal`, `lowerAlpha`, `lowerRoman`.
     */
    schemes?: readonly EditorOrderedListNumberingScheme[];
    /** Punctuation drawn after the formatted index. Defaults to `.`. */
    suffix?: '.' | ')';
}

/** Layout of list indentation and markers. */
export interface EditorListTheme {
    /** Indentation added per nesting depth, in layout units. */
    indent?: number;
    /** Multiplier applied to the indentation of the outermost list level. */
    baseIndentMultiplier?: number;
    /** Space between consecutive list items, in layout units. */
    itemSpacing?: number;
    /** Space after a list before following content, including nested lists, in layout units. */
    spacingAfter?: number;
    /** Color of bullets and numbers. Defaults to the resolved text color. */
    markerColor?: string;
    /** Scale applied to the bullet glyph of unordered lists, relative to the item's font size. */
    markerScale?: number;
    /** Gap between the marker and the item text, in layout units. */
    markerGap?: number;
    /** Ordered-list marker presentation by visual nesting depth. */
    orderedMarker?: EditorOrderedListMarkerTheme;
}

/** The rule drawn for `horizontalRule` nodes. */
export interface EditorHorizontalRuleTheme {
    color?: string;
    /** Rule line thickness, in layout units. */
    thickness?: number;
    /** Space above and below the rule, in layout units. */
    verticalMargin?: number;
}

/** Blockquote typography and the leading border bar. */
export interface EditorBlockquoteTheme {
    /** Typography for text inside a blockquote, merged over the base `text` style. */
    text?: EditorTextStyle;
    /** Leading inset reserved per blockquote depth, in layout units. */
    indent?: number;
    /** Color of the vertical bar drawn beside the quote. */
    borderColor?: string;
    /** Width of the vertical bar, in layout units. */
    borderWidth?: number;
    /** Gap between the bar and the quoted text, in layout units. */
    markerGap?: number;
}

/** Code block typography and its background panel. */
export interface EditorCodeBlockTheme {
    /** Typography for code text. Falls back to the platform monospace font. */
    text?: EditorTextStyle;
    /** Fill drawn behind the block. */
    backgroundColor?: string;
    /** Panel corner radius, in layout units. */
    borderRadius?: number;
    /** Inner horizontal padding, in layout units. */
    paddingHorizontal?: number;
    /** Inner vertical padding, in layout units. */
    paddingVertical?: number;
}

/**
 * Toolbar chrome style. `'custom'` renders the package's own chrome from
 * this theme; `'native'` adopts the platform keyboard-accessory look, so the
 * platform supplies chrome metrics this theme would otherwise set.
 */
export type EditorToolbarAppearance = 'custom' | 'native';

/**
 * Toolbar chrome and button colors. Honored by both the JavaScript
 * `EditorToolbar` and the native keyboard toolbar unless a field says
 * otherwise.
 */
export interface EditorToolbarTheme {
    /** Chrome style. Defaults to `'custom'`. */
    appearance?: EditorToolbarAppearance;
    /** Toolbar height, in layout units. Defaults to a height that fits the buttons. */
    height?: number;
    backgroundColor?: string;
    borderColor?: string;
    /** Toolbar border width, in layout units. */
    borderWidth?: number;
    /** Toolbar corner radius, in layout units. */
    borderRadius?: number;
    /** Gap above an inline-placed toolbar, in layout units. Defaults to 8.
     *  Ignored for `toolbarPlacement="keyboard"`. */
    marginTop?: number;
    /** Whether the JavaScript toolbar draws its top separator line. Defaults to
     *  false through `NativeRichTextEditor`, true for a standalone `EditorToolbar`. */
    showTopBorder?: boolean;
    /** Gap between the keyboard and the toolbar, in layout units. Applies to
     *  `toolbarPlacement="keyboard"`; the platform default depends on `appearance`. */
    keyboardOffset?: number;
    /** Inset on each side of the keyboard-attached toolbar, in layout units.
     *  The platform default depends on `appearance`. */
    horizontalInset?: number;
    /** Color of `separator` toolbar items. */
    separatorColor?: string;
    /** Icon color of an idle button. */
    buttonColor?: string;
    /** Icon size for toolbar buttons, in layout units. */
    buttonIconSize?: number;
    /** Icon color of a button whose mark or node is active at the selection. */
    buttonActiveColor?: string;
    /** Icon color of a button the current selection cannot apply. */
    buttonDisabledColor?: string;
    /** Fill drawn behind an active button. */
    buttonActiveBackgroundColor?: string;
    /** Button corner radius, in layout units. */
    buttonBorderRadius?: number;
}

/** Padding between the editor's edges and its content, in layout units. */
export interface EditorContentInsets {
    top?: number;
    right?: number;
    bottom?: number;
    left?: number;
}

/**
 * Native content theme for `NativeRichTextEditor` and `NativeProseViewer`.
 *
 * Numeric values are layout units — density-independent pixels on Android,
 * points on iOS. Colors are strings, and the portable formats are `#RGB`,
 * `#RGBA`, `#RRGGBB`, `#RRGGBBAA`, `rgb(r, g, b)`, `rgba(r, g, b, a)`, and
 * `transparent`; named colors beyond `black`, `white`, `red`, `green`,
 * `blue`, and `gray` resolve on Android only.
 *
 * Text styles cascade: `text` is the base every block inherits, then
 * `blockquote.text` (inside a blockquote), then the node's own entry —
 * `paragraph`, `codeBlock.text`, or `headings.h1`…`headings.h6`. One
 * exception: `text.lineHeight` does not reach paragraphs, so set
 * `paragraph.lineHeight` to give paragraphs a line height.
 *
 * Mention styling lives on the mentions addon, not here — see
 * {@link EditorMentionTheme}.
 */
export interface EditorTheme {
    /** Base typography inherited by every block. */
    text?: EditorTextStyle;
    /** Paragraph typography, merged over `text`. */
    paragraph?: EditorTextStyle;
    blockquote?: EditorBlockquoteTheme;
    codeBlock?: EditorCodeBlockTheme;
    headings?: EditorHeadingTheme;
    list?: EditorListTheme;
    horizontalRule?: EditorHorizontalRuleTheme;
    links?: EditorLinkTheme;
    toolbar?: EditorToolbarTheme;
    /** Color of the editor's placeholder text. */
    placeholderColor?: string;
    /** Fill drawn behind the content. */
    backgroundColor?: string;
    /** Corner radius of the content surface, in layout units. */
    borderRadius?: number;
    contentInsets?: EditorContentInsets;
}

function stripUndefined(value: unknown): unknown {
    if (Array.isArray(value)) {
        return value.map((item) => stripUndefined(item)).filter((item) => item !== undefined);
    }

    if (value != null && typeof value === 'object') {
        const entries = Object.entries(value as Record<string, unknown>)
            .map(([key, entryValue]) => [key, stripUndefined(entryValue)] as const)
            .filter(([, entryValue]) => entryValue !== undefined);
        if (entries.length === 0) {
            return undefined;
        }
        return Object.fromEntries(entries);
    }

    if (typeof value === 'number' && !Number.isFinite(value)) {
        return undefined;
    }

    return value;
}

// Mentions are configured on the mentions addon, not on EditorTheme; native
// still reads them from the theme payload.
export function serializeEditorTheme(
    theme?: EditorTheme,
    mentionTheme?: EditorMentionTheme
): string | undefined {
    const cleanedTheme = theme ? stripUndefined(theme) : undefined;
    const base =
        cleanedTheme && typeof cleanedTheme === 'object'
            ? (cleanedTheme as Record<string, unknown>)
            : undefined;
    const cleanedMentions = mentionTheme ? stripUndefined(mentionTheme) : undefined;

    if (cleanedMentions == null) {
        return base ? JSON.stringify(base) : undefined;
    }
    return JSON.stringify({ ...(base ?? {}), mentions: cleanedMentions });
}
